File: MiszIQ/MiszIQ/AudioManager.swift
import AVFoundation
import SwiftUI

enum SoundEffect: String, CaseIterable {
    case correctAnswer = "correct_chime"
    case wrongAnswer = "wrong_buzz"
    case buttonTap = "button_tap"
    case gameComplete = "celebration"
}

class AudioManager: ObservableObject {
    static let shared = AudioManager()

    private var musicPlayer: AVAudioPlayer?
    private var soundPlayers: [SoundEffect: AVAudioPlayer] = [:]
    private var isMusicPlaying = false

    private init() {
        setupAudioSession()
        preloadSoundEffects()
    }

    // MARK: - Setup

    private func setupAudioSession() {
        do {
            // Use ambient category so music mixes with other audio and respects silent mode
            try AVAudioSession.sharedInstance().setCategory(.ambient, mode: .default)
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
            print("Failed to setup audio session: \(error)")
        }
    }

    private func preloadSoundEffects() {
        for effect in SoundEffect.allCases {
            if let url = Bundle.main.url(forResource: effect.rawValue, withExtension: "mp3") {
                do {
                    let player = try AVAudioPlayer(contentsOf: url)
                    player.prepareToPlay()
                    soundPlayers[effect] = player
                } catch {
                    print("Failed to preload \(effect.rawValue): \(error)")
                }
            }
        }
    }

    // MARK: - Background Music

    func playBackgroundMusic() {
        guard SettingsManager.shared.musicEnabled else { return }
        guard !isMusicPlaying else { return }

        // Try to load background music
        if let url = Bundle.main.url(forResource: "background_music", withExtension: "mp3") {
            do {
                musicPlayer = try AVAudioPlayer(contentsOf: url)
                musicPlayer?.numberOfLoops = -1 // Infinite loop
                musicPlayer?.volume = 0.3 // 30% volume
                musicPlayer?.play()
                isMusicPlaying = true
            } catch {
                print("Failed to play background music: \(error)")
            }
        }
    }

    func stopBackgroundMusic() {
        musicPlayer?.stop()
        musicPlayer?.currentTime = 0
        isMusicPlaying = false
    }

    func pauseBackgroundMusic() {
        musicPlayer?.pause()
    }

    func resumeBackgroundMusic() {
        guard SettingsManager.shared.musicEnabled else { return }
        guard isMusicPlaying else { return }
        musicPlayer?.play()
    }

    // MARK: - Sound Effects

    func playSoundEffect(_ effect: SoundEffect) {
        guard SettingsManager.shared.soundEffectsEnabled else { return }

        // Try preloaded player first
        if let player = soundPlayers[effect] {
            player.currentTime = 0
            player.play()
            return
        }

        // Fallback: create new player
        guard let url = Bundle.main.url(forResource: effect.rawValue, withExtension: "mp3") else {
            return
        }

        do {
            let player = try AVAudioPlayer(contentsOf: url)
            player.play()
            soundPlayers[effect] = player
        } catch {
            print("Failed to play sound effect \(effect.rawValue): \(error)")
        }
    }

    // MARK: - Convenience Methods

    func playCorrectSound() {
        playSoundEffect(.correctAnswer)
    }

    func playWrongSound() {
        playSoundEffect(.wrongAnswer)
    }

    func playButtonTap() {
        playSoundEffect(.buttonTap)
    }

    func playGameComplete() {
        playSoundEffect(.gameComplete)
    }
}

File: MiszIQAndroid/app/src/main/java/com/example/misziq/audio/AudioManager.kt
package com.example.misziq.audio

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaPlayer
import android.media.SoundPool
import com.example.misziq.R
import com.example.misziq.data.preferences.SettingsDataStore
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking

enum class SoundEffect(val resourceName: String) {
    CORRECT_ANSWER("correct_chime"),
    WRONG_ANSWER("wrong_buzz"),
    BUTTON_TAP("button_tap"),
    GAME_COMPLETE("celebration")
}

class AudioManager(private val context: Context) {
    private var mediaPlayer: MediaPlayer? = null
    private var soundPool: SoundPool? = null
    private val soundIds = mutableMapOf<SoundEffect, Int>()
    private val settingsDataStore = SettingsDataStore(context)
    private var isMusicPlaying = false

    fun initialize() {
        // Initialize SoundPool for sound effects
        val audioAttributes = AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_GAME)
            .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
            .build()

        soundPool = SoundPool.Builder()
            .setMaxStreams(4)
            .setAudioAttributes(audioAttributes)
            .build()

        // Load sound effects
        loadSoundEffect(SoundEffect.CORRECT_ANSWER, "correct_chime")
        loadSoundEffect(SoundEffect.WRONG_ANSWER, "wrong_buzz")
        loadSoundEffect(SoundEffect.BUTTON_TAP, "button_tap")
        loadSoundEffect(SoundEffect.GAME_COMPLETE, "celebration")
    }

    private fun loadSoundEffect(effect: SoundEffect, resourceName: String) {
        val resId = context.resources.getIdentifier(resourceName, "raw", context.packageName)
        if (resId != 0) {
            soundPool?.load(context, resId, 1)?.let { id ->
                soundIds[effect] = id
            }
        }
    }

    // MARK: - Background Music

    fun playBackgroundMusic() {
        val musicEnabled = runBlocking { settingsDataStore.musicEnabled.first() }
        if (!musicEnabled) return
        if (isMusicPlaying) return

        val resId = context.resources.getIdentifier("background_music", "raw", context.packageName)
        if (resId != 0) {
            try {
                mediaPlayer = MediaPlayer.create(context, resId)?.apply {
                    isLooping = true
                    setVolume(0.3f, 0.3f)
                    start()
                }
                isMusicPlaying = true
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun stopBackgroundMusic() {
        mediaPlayer?.stop()
        mediaPlayer?.release()
        mediaPlayer = null
        isMusicPlaying = false
    }

    fun pauseBackgroundMusic() {
        mediaPlayer?.pause()
    }

    fun resumeBackgroundMusic() {
        val musicEnabled = runBlocking { settingsDataStore.musicEnabled.first() }
        if (!musicEnabled) return
        if (!isMusicPlaying) return
        mediaPlayer?.start()
    }

    // MARK: - Sound Effects

    fun playSoundEffect(effect: SoundEffect) {
        val soundEnabled = runBlocking { settingsDataStore.soundEffectsEnabled.first() }
        if (!soundEnabled) return

        soundIds[effect]?.let { soundId ->
            soundPool?.play(soundId, 1f, 1f, 1, 0, 1f)
        }
    }

    fun playCorrectSound() = playSoundEffect(SoundEffect.CORRECT_ANSWER)
    fun playWrongSound() = playSoundEffect(SoundEffect.WRONG_ANSWER)
    fun playButtonTap() = playSoundEffect(SoundEffect.BUTTON_TAP)
    fun playGameComplete() = playSoundEffect(SoundEffect.GAME_COMPLETE)

    // MARK: - Cleanup

    fun release() {
        mediaPlayer?.release()
        mediaPlayer = null
        soundPool?.release()
        soundPool = null
        isMusicPlaying = false
    }
}
